include/oneapi/dpl/internal/async_impl/async_impl.h

```diff
@@ -0,0 +1,89 @@
+// -*- C++ -*-
+//===----------------------------------------------------------------------===//
+//
+// Copyright (C) 2021 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+//
+// This file incorporates work covered by the following copyright and permission
+// notice:
+//
+// Part of the LLVM Project, under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef _ONEDPL_ASYNC_IMPL_H
+#define _ONEDPL_ASYNC_IMPL_H
+
+#if _ONEDPL_HETERO_BACKEND
+#    include "async_impl_hetero.h"
+#endif
+
+#include "glue_async_impl.h"
+
+namespace oneapi
+{
+namespace dpl
+{
+namespace experimental
+{
+
+// [wait_for_all]
+template <typename... _Ts>
+void
+wait_for_all(const _Ts&... __Events)
+{
+    ::std::vector<sycl::event> __wait_list = {__Events...};
+    for (auto _a : __wait_list)
```

Comment:
You can even avoid the loop and the copying of the event with https://godbolt.org/z/zK4a1q. Please have a look at `foo` function body.

You may need to get rid of `const` or probably it's not a problem if `event` is semantically shared pointer.